README.md
# ScummVM libretro Core

[ScummVM](http://scummvm.org) is an interpreter for point-and-click adventure games that can be used as a libretro core.

## Compiling

To compile the core, run the following:

```
cd backends/platform/libretro/build
make
```

---

# [ScummVM README](https://www.scummvm.org/) · [![Build Status](https://travis-ci.org/scummvm/scummvm.svg?branch=master)](https://travis-ci.org/scummvm/scummvm) [![Translation status](https://translations.scummvm.org/widgets/scummvm/-/scummvm/svg-badge.svg)](https://translations.scummvm.org/engage/scummvm/?utm_source=widget) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](CONTRIBUTING.md#pull-requests) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/200a9bd3f7a647b48415efb484cc8bdc)](https://www.codacy.com/app/sev-/scummvm?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=scummvm/scummvm&amp;utm_campaign=Badge_Grade)

For more information, compatibility lists, details on donating, the
latest release, progress reports and more, please visit the ScummVM home
page at: <https://www.scummvm.org/>

## Table of Contents:

  - [1.0) Introduction](#10-introduction)
      - [1.1) About ScummVM](#11-about-scummvm)
      - [1.2) Quick start](#12-quick-start)
      - [1.3) F.A.Q.](#13-faq)
  - [2.0) Contact](#20-contact)
      - [2.1) Reporting Bugs](#21-reporting-bugs)
  - [3.0) Supported Games](#30-supported-games)
      - [3.1) Copy Protection](#31-copy-protection)
      - [3.2) Datafiles](#32-datafiles)
      - [3.3) Multi-CD games notes](#33-multi-cd-games-notes)
      - [3.4) Known Problems](#34-known-problems)
      - [3.5) Extra Data Files](#35-extra-data-files)
      - [3.6) Blade Runner notes](#36-blade-runner-notes)
      - [3.7) Broken Sword games notes](#37-broken-sword-games-notes)
          - [3.7.1) Broken Sword](#371-broken-sword)
          - [3.7.2) Broken Sword II](#372-broken-sword-ii)
          - [3.7.3) Broken Sword games
            cutscenes](#373-broken-sword-games-cutscenes)
          - [3.7.4) Broken Sword games cutscenes, in
            retrospect](#374-broken-sword-games-cutscenes-in-retrospect)
      - [3.8) Day of the Tentacle notes](#38-day-of-the-tentacle-notes)
      - [3.9) Dragon History notes](#39-dragon-history-notes)
      - [3.10) Flight of the Amazon Queen
        notes](#310-flight-of-the-amazon-queen-notes)
      - [3.11) Gobliiins notes](#311-gobliiins-notes)
      - [3.12) Inherit the Earth: Quest for the Orb
        notes](#312-inherit-the-earth-quest-for-the-orb-notes)
      - [3.13) Mickey's Space Adventure
        notes](#313-mickeys-space-adventure-notes)
      - [3.14) Might and Magic Xeen games
        notes](#314-might-and-magic-xeen-games-notes)
      - [3.15) Myst game notes](#315-myst-game-notes)
      - [3.16) Quest for Glory notes](#316-quest-for-glory-notes)
      - [3.17) Riven game notes](#317-riven-game-notes)
      - [3.18) Simon the Sorcerer games
        notes](#318-simon-the-sorcerer-games-notes)
      - [3.19) Starship Titanic game
        notes](#319-starship-titanic-game-notes)
      - [3.20) The Curse of Monkey Island
        notes](#320-the-curse-of-monkey-island-notes)
      - [3.21) The Feeble Files notes](#321-the-feeble-files-notes)
      - [3.22) The Legend of Kyrandia
        notes](#322-the-legend-of-kyrandia-notes)
      - [3.23) Troll's Tale notes](#323-trolls-tale-notes)
      - [3.24) Winnie the Pooh notes](#324-winnie-the-pooh-notes)
      - [3.25) Sierra AGI games: Predictive Input
        Dialog](#325-sierra-agi-games-predictive-input-dialog)
      - [3.26) Sierra SCI games: Simultaneous speech and
        subtitles](#326-sierra-sci-games-simultaneous-speech-and-subtitles)
      - [3.27) Zork games notes](#327-zork-games-notes)
          - [3.27.1) Zork Nemesis: The Forbidden
            Lands](#3271-zork-nemesis-the-forbidden-lands)
          - [3.27.2) Zork: Grand
            Inquisitor](#3272-zork-grand-inquisitor)
      - [3.28) Commodore64 games notes](#328-commodore64-games-notes)
      - [3.29) Macintosh games notes](#329-macintosh-games-notes)
  - [4.0) Supported Platforms](#40-supported-platforms)
  - [5.0) Running ScummVM](#50-running-scummvm)
      - [5.1) Command Line Options](#51-command-line-options)
      - [5.2) Global Menu](#52-global-menu)
      - [5.3) Graphics filters](#53-graphics-filters)
      - [5.4) Hotkeys](#54-hotkeys)
      - [5.5) Language options](#55-language-options)
  - [6.0) Saved Games](#60-saved-games)
      - [6.1) Autosaves](#61-autosaves)
      - [6.2) Converting Saved Games](#62-converting-saved-games)
      - [6.3) Viewing/Loading saved games from the command
        line](#63-viewingloading-saved-games-from-the-command-line)
  - [7.0) Music and Sound](#70-music-and-sound)
      - [7.1) AdLib emulation](#71-adlib-emulation)
      - [7.2) FluidSynth MIDI emulation](#72-fluidsynth-midi-emulation)
      - [7.3) MT-32 emulation](#73-mt-32-emulation)
      - [7.4) MIDI emulation](#74-midi-emulation)
      - [7.5) Native MIDI support](#75-native-midi-support)
          - [7.5.1) Using MIDI options to customize Native MIDI
            output](#751-using-midi-options-to-customize-native-midi-output)
      - [7.6) UNIX native, ALSA and dmedia sequencer
        support](#76-unix-native-alsa-and-dmedia-sequencer-support)
          - [7.6.1) ALSA sequencer \[UNIX
            ONLY\]](#761-alsa-sequencer-unix-only)
          - [7.6.2) IRIX dmedia sequencer: \[UNIX
            ONLY\]](#762-irix-dmedia-sequencer-unix-only)
      - [7.7) TiMidity++ MIDI server
        support](#77-timidity-midi-server-support)
      - [7.8) Using compressed audio
        files](#78-using-compressed-audio-files)
          - [7.8.1) Using MP3 files for CD
            audio](#781-using-mp3-files-for-cd-audio)
          - [7.8.2) Using Ogg Vorbis files for CD
            audio](#782-using-ogg-vorbis-files-for-cd-audio)
          - [7.8.3) Using Flac files for CD
            audio](#783-using-flac-files-for-cd-audio)
          - [7.8.4) Compressing MONSTER.SOU with
            MP3](#784-compressing-monstersou-with-mp3)
          - [7.8.5) Compressing MONSTER.SOU with Ogg
            Vorbis](#785-compressing-monstersou-with-ogg-vorbis)
          - [7.8.6) Compressing MONSTER.SOU with
            Flac](#786-compressing-monstersou-with-flac)
          - [7.8.7) Compressing music/sfx/speech in AGOS
            games](#787-compressing-musicsfxspeech-in-agos-games)
          - [7.8.8) Compressing speech/music in Broken
            Sword](#788-compressing-speechmusic-in-broken-sword)
          - [7.8.9) Compressing speech/music in Broken Sword
            II](#789-compressing-speechmusic-in-broken-sword-ii)
      - [7.9) Output sample rate](#79-output-sample-rate)
  - [8.0) Configuration file](#80-configuration-file)
      - [8.1) Recognized configuration
        keywords](#81-recognized-configuration-keywords)
      - [8.2) Custom game options that can be toggled via the
        GUI](#82-custom-game-options-that-can-be-toggled-via-the-gui)
  - [9.0) Screenshots (SDL backend
    only)](#90-screenshots-sdl-backend-only)
  - [10.0) Compiling](#100-compiling)
  - [11.0) Changelog](#110-changelog)
  - [12.0) Credits](#120-credits)

## 1.0) Introduction

### 1.1) About ScummVM

ScummVM is a program which allows you to run certain classic graphical
point-and-click adventure games, provided you already have their data
files. The clever part about this: ScummVM just replaces the executables
shipped with the game, allowing you to play them on systems for which
they were never designed\!

Originally it was designed to run LucasArts' SCUMM games, such as Maniac
Mansion, Monkey Island, Day of the Tentacle or Sam and Max. SCUMM stands
for 'Script Creation Utility for Maniac Mansion', which was the first
game for which LucasArts designed this system. And much later it gave
its name to ScummVM ('VM' meaning Virtual Machine).

Over time support for a lot of non-SCUMM games has been added, and
ScummVM now also supports many of Sierra's AGI and SCI games (such as
King's Quest 1-6, Space Quest 1-5, ...), Discworld 1 and 2, Simon the
Sorcerer 1 and 2, Beneath A Steel Sky, Lure of the Temptress, Broken
Sword I and II, Flight of the Amazon Queen, Gobliiins 1-3, The Legend of
Kyrandia series, many of Humongous Entertainment's children's SCUMM
games (including Freddi Fish and Putt Putt games) and many more. You can
find a full list with details on which adventures are supported and how
well on the compatibility page. ScummVM is continually improving, so
check back often.

Among the systems on which you can play those games are regular desktop
computers (running Windows, Linux, Mac OS X, ...), game consoles
(Dreamcast, Nintendo DS & Wii, PS2, PSP, ...), smartphones (Android,
iPhone, PocketPC, Symbian ...) and more.

At this time ScummVM is still under heavy development. Be aware that
whilst we attempt to make sure that many games can be completed with few
major bugs, crashes can happen and we offer no warranty. That being
said, some of the games have been supported for a long time and should
work fine with any recent stable release. You can get a feeling of how
well each game is working in ScummVM by looking at the compatibility
page. Actually, if you browse a bit around you might discover that
ScummVM is even being used commercially to re-release some of the
supported games on modern platforms. This shows that several companies
are happy with the quality of the software and how well it can run some
of the games.

If you enjoy ScummVM feel free to donate using the PayPal button on the
ScummVM homepage. This will help us buy utilities needed to develop
ScummVM easier and quicker. If you cannot donate, help and contribute a
patch\!

### 1.2) Quick start


NEWS.md
For a more comprehensive changelog of the latest experimental code, see:
        https://github.com/scummvm/scummvm/commits/

#### 2.1.0 (2019-10-13)

 New games:
   - Added support for Blade Runner.
   - Added support for Duckman: The Graphic Adventures of a Private Dick.
   - Added support for Hoyle Bridge.
   - Added support for Hoyle Children's Collection.
   - Added support for Hoyle Classic Games.
   - Added support for Hoyle Solitaire.
   - Added support for Hyperspace Delivery Boy!
   - Added support for Might and Magic IV - Clouds of Xeen.
   - Added support for Might and Magic V - Darkside of Xeen.
   - Added support for Might and Magic - World of Xeen.
   - Added support for Might and Magic - World of Xeen 2 CD Talkie.
   - Added support for Might and Magic - Swords of Xeen.
   - Added support for Mission Supernova Part 1.
   - Added support for Mission Supernova Part 2.
   - Added support for Quest for Glory: Shadows of Darkness.
   - Added support for The Prince and the Coward.
   - Added support for Versailles 1685.

 New ports:
   - Added Nintendo Switch port.

 General:
   - Improved GUI rendering and overall GUI performance.
   - Added stretch mode option to control how the display is stretched to the
     window or screen area.
   - Fixed incorrect cursor movement when it's controlled using the keyboard.
   - Updated the Roland MT-32 emulation code to Munt 2.3.0.
   - Improved unknown game variants reporting.
   - Enabled cloud support.
   - Added Text to Speech capabilities for better accessibility on some platforms.

 ADL:
   - Improved color accuracy.
   - Added a TV emulation mode.
   - Added support for the WOZ disk image format.

 Drascula:
   - Fixed loading game from launcher when the game had been saved in chapter 1.

 Full Pipe:
   - Fixed playtime not being restored when loading a savegame.
   - Fixed a bug that leads to enormous memory consumption in scene 22.

 Kyra:
   - Added support for the Amiga version of Eye of the Beholder I + II.
   - Added support for the FM-Towns version of Eye of the Beholder II.
   - Several bug fixes.

 MOHAWK:
   - Added a main menu for the 25th anniversary release of Myst ME.
   - Repurposed the landing menu as a main menu for the 25th anniversary
     release of Riven.
   - Added autosave to slot 0 to Myst and Riven.
   - Added keyboard shortcuts for loading and saving as documented in the game
     manual to Myst and Riven.
   - Fixed a crash caused by the observatory viewer random position going out
     of bounds in Myst.
   - Fixed a crash caused by Jungle Island flies going out of bounds in Riven.
   - Fixed missing end credits for the Polish version in Riven.
   - Improved usability for some puzzles in Myst and Riven.
   - Fixed various crashes, graphics glitches, and sound imperfections in Myst
     and Riven.

 Mortevielle:
   - Added speech synthesis on some platforms.

 SCI:
   - Added LarryScale, a high quality cartoon scaler for Leisure Suit Larry 7.
   - Fixed over 100 original game script bugs in CAMELOT, ECO1, ECO2,
     FREDDYPHARKAS, GK1, HOYLE5, ICEMAN, KQ6, LB1, LB2, LONGBOW, LSL6,
     MOTHERGOOSE256, PQ3, PQ4, QFG1VGA, QFG4, and SQ4.
   - Fixed a bug in version 2.0.0 that prevented the Macintosh versions of
     Freddy Pharkas, King's Quest 6, and Quest for Glory 1 (VGA) from loading.
   - Fixed a crash in the Macintosh version of Freddy Pharkas when picking up
     the shovel which makes the game completable.
   - Fixed loading autosave games.

 SCUMM:
   - Implemented lipsync for v6 and v7+ games.
   - Improved Audio quality in Humongous Entertainment games by using the Miles AdLib driver.
   - Fixed possible stack overflows in The Dig and Full Throttle.
   - Fixed original speech glitch on submarine in Indiana Jones and the Fate of Atlantis.
     Users need to recompress their monster.sou using an up-to-date version of scummvm-tools
     for this to take effect when using compressed audio.
   - Fixed an issue in the wig maker room in the German version of SPY Fox 3: Operation Ozone
     which makes the game completable.
   - Added sound driver for the Amiga versions of Monkey Island 2 and Indiana Jones
     and the Fate of Atlantis.

 Sherlock:
   - Fixed crash in Spanish version talking to lady in Tailor shop.

 SKY:
   - Added workaround for original game bug to improve intro and not cut off images which exist
     as fullscreen (320x200px) in the game data files.

 Tinsel:
   - Fix loading Discworld 1 savegames from the launcher where Rincewind had a held item.
   - Script patch for hang in Discworld 1 GRA using items on Temple big hammer.
   - In Discworld 1, Held items being released that were never in the Luggage or Rincewind's inventory
     will now be automatically dropped into the Luggage rather than being lost.

 Titanic:
   - Fixed bug in entering floor numbers numerically that could crash the game.

engines/queen/sound.cpp
/* ScummVM - Graphic Adventure Engine
 *
 * ScummVM is the legal property of its developers, whose names
 * are too numerous to list here. Please refer to the COPYRIGHT
 * file distributed with this source distribution.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 */


#include "common/config-manager.h"
#include "common/endian.h"
#include "common/memstream.h"
#include "common/textconsole.h"

#include "queen/sound.h"
#include "queen/input.h"
#include "queen/logic.h"
#include "queen/music.h"
#include "queen/queen.h"
#include "queen/resource.h"

#include "audio/audiostream.h"
#include "audio/decoders/flac.h"
#include "audio/decoders/mp3.h"
#include "audio/decoders/raw.h"
#include "audio/decoders/vorbis.h"
#include "audio/mods/rjp1.h"

#define	SB_HEADER_SIZE_V104 110
#define	SB_HEADER_SIZE_V110 122

namespace Queen {

// The sounds in the PC versions are all played at 11840 Hz. Unfortunately, we
// did not know that at the time, so there are plenty of compressed versions
// which claim that they should be played at 11025 Hz. This "wrapper" class
// works around that.

class AudioStreamWrapper : public Audio::AudioStream {
protected:
	Audio::AudioStream *_stream;
	int _rate;

public:
	AudioStreamWrapper(Audio::AudioStream *stream) {
		_stream = stream;

		int rate = _stream->getRate();

		// A file where the sample rate claims to be 11025 Hz is
		// probably compressed with the old tool. We force the real
		// sample rate, which is 11840 Hz.
		//
		// However, a file compressed with the newer tool is not
		// guaranteed to have a sample rate of 11840 Hz. LAME will
		// automatically resample it to 12000 Hz. So in all other
		// cases, we use the rate from the file.

		if (rate == 11025)
			_rate = 11840;
		else
			_rate = rate;
	}
	~AudioStreamWrapper() {
		delete _stream;
	}
	int readBuffer(int16 *buffer, const int numSamples) {
		return _stream->readBuffer(buffer, numSamples);
	}
	bool isStereo() const {
		return _stream->isStereo();
	}
	bool endOfData() const {
		return _stream->endOfData();
	}
	bool endOfStream() const {
		return _stream->endOfStream();
	}
	int getRate() const {
		return _rate;
	}
};

class SilentSound : public PCSound {
public:
	SilentSound(Audio::Mixer *mixer, QueenEngine *vm) : PCSound(mixer, vm) {}
protected:
	void playSoundData(Common::File *f, uint32 size, Audio::SoundHandle *soundHandle) {
		// Do nothing
	}
};

class SBSound : public PCSound {
public:
	SBSound(Audio::Mixer *mixer, QueenEngine *vm) : PCSound(mixer, vm) {}
protected:
	void playSoundData(Common::File *f, uint32 size, Audio::SoundHandle *soundHandle);
};

#ifdef USE_MAD
class MP3Sound : public PCSound {
public:
	MP3Sound(Audio::Mixer *mixer, QueenEngine *vm) : PCSound(mixer, vm) {}
protected:
	void playSoundData(Common::File *f, uint32 size, Audio::SoundHandle *soundHandle) {
		Common::SeekableReadStream *tmp = f->readStream(size);
		assert(tmp);
		_mixer->playStream(Audio::Mixer::kSFXSoundType, soundHandle, new AudioStreamWrapper(Audio::makeMP3Stream(tmp, DisposeAfterUse::YES)));
	}
};
#endif

#ifdef USE_VORBIS
class OGGSound : public PCSound {
public:
	OGGSound(Audio::Mixer *mixer, QueenEngine *vm) : PCSound(mixer, vm) {}
protected:
	void playSoundData(Common::File *f, uint32 size, Audio::SoundHandle *soundHandle) {
		Common::SeekableReadStream *tmp = f->readStream(size);
		assert(tmp);
		_mixer->playStream(Audio::Mixer::kSFXSoundType, soundHandle, new AudioStreamWrapper(Audio::makeVorbisStream(tmp, DisposeAfterUse::YES)));
	}
};
#endif

#ifdef USE_FLAC
class FLACSound : public PCSound {
public:
	FLACSound(Audio::Mixer *mixer, QueenEngine *vm) : PCSound(mixer, vm) {}
protected:
	void playSoundData(Common::File *f, uint32 size, Audio::SoundHandle *soundHandle) {
		Common::SeekableReadStream *tmp = f->readStream(size);
		assert(tmp);
		_mixer->playStream(Audio::Mixer::kSFXSoundType, soundHandle, new AudioStreamWrapper(Audio::makeFLACStream(tmp, DisposeAfterUse::YES)));
	}
};
#endif // #ifdef USE_FLAC

Sound::Sound(Audio::Mixer *mixer, QueenEngine *vm) :
	_mixer(mixer), _vm(vm), _sfxToggle(true), _speechToggle(true), _musicToggle(true),
	_speechSfxExists(false), _lastOverride(0), _musicVolume(0) {
}

Sound *Sound::makeSoundInstance(Audio::Mixer *mixer, QueenEngine *vm, uint8 compression) {
	if (vm->resource()->getPlatform() == Common::kPlatformAmiga)
		return new AmigaSound(mixer, vm);

	switch (compression) {
	case COMPRESSION_NONE:
		return new SBSound(mixer, vm);
	case COMPRESSION_MP3:
#ifndef USE_MAD
		warning("Using MP3 compressed datafile, but MP3 support not compiled in");
		return new SilentSound(mixer, vm);
#else
		return new MP3Sound(mixer, vm);
#endif
	case COMPRESSION_OGG:
#ifndef USE_VORBIS
		warning("Using OGG compressed datafile, but OGG support not compiled in");
		return new SilentSound(mixer, vm);
#else
		return new OGGSound(mixer, vm);
#endif
	case COMPRESSION_FLAC:
#ifndef USE_FLAC
		warning("Using FLAC compressed datafile, but FLAC support not compiled in");
		return new SilentSound(mixer, vm);
#else
		return new FLACSound(mixer, vm);
#endif
	default:
		warning("Unknown compression type");
		return new SilentSound(mixer, vm);
	}
}

void Sound::setVolume(int vol) {
	if (ConfMan.hasKey("mute") && ConfMan.getBool("mute"))
		_musicVolume = 0;
	else
		_musicVolume = vol;

	_mixer->setVolumeForSoundType(Audio::Mixer::kMusicSoundType, _musicVolume);
}

void Sound::saveState(byte *&ptr) {
	WRITE_BE_UINT16(ptr, _lastOverride); ptr += 2;
}

void Sound::loadState(uint32 ver, byte *&ptr) {
	_lastOverride = (int16)READ_BE_INT16(ptr); ptr += 2;
}

PCSound::PCSound(Audio::Mixer *mixer, QueenEngine *vm)
	: Sound(mixer, vm) {

	_music = new MidiMusic(vm);
}

PCSound::~PCSound() {
	delete _music;
}

void PCSound::playSfx(uint16 sfx) {
	if (sfxOn() && sfx != 0)
		playSound(_sfxName[sfx - 1], false);
}

void PCSound::playSong(int16 songNum) {
	if (songNum <= 0) {
		_music->stopSong();
		return;
	}

	int16 newTune;
	if (_vm->resource()->isDemo()) {
		if (songNum == 17) {
			_music->stopSong();
			return;
		}
		newTune = _songDemo[songNum - 1].tuneList[0] - 1;
	} else {
		newTune = _song[songNum - 1].tuneList[0] - 1;
	}

	if (_tune[newTune].sfx[0]) {
		playSfx(_tune[newTune].sfx[0]);
		return;
	}

	if (!musicOn())
		return;

	int overrideCmd = (_vm->resource()->isDemo()) ? _songDemo[songNum - 1].overrideCmd : _song[songNum - 1].overrideCmd;
	switch (overrideCmd) {
	// Override all songs
	case  1:
		break;
	// Alter song settings (such as volume) and exit
	case  2:
		_music->toggleVChange();
	default:
		return;
	}

	_lastOverride = songNum;

	_music->queueTuneList(newTune);
	_music->playMusic();
}

void PCSound::stopSong() {
	_music->stopSong();
}

void PCSound::playSpeech(const char *base) {
	if (speechOn()) {
		playSound(base, true);
	}
}

void PCSound::setVolume(int vol) {
	Sound::setVolume(vol);
	_music->setVolume(vol);
}

void PCSound::playSound(const char *base, bool isSpeech) {
	char name[13];
	strcpy(name, base);
	// alter filename to add zeros and append ".SB"
	for (int i = 0; i < 8; i++) {
		if (name[i] == ' ')
			name[i] = '0';
	}
	strcat(name, ".SB");
	if (isSpeech) {
#ifdef __LIBRETRO__
		// Add _vm->shouldQuit() check here, otherwise game gets stuck
		// in an infinite loop if we try to quit while a sound is playing...
		while (_mixer->isSoundHandleActive(_speechHandle) && !_vm->shouldQuit()) {
#else
		while (_mixer->isSoundHandleActive(_speechHandle)) {
#endif
			_vm->input()->delay(10);
		}
	} else {
		_mixer->stopHandle(_sfxHandle);
	}
	uint32 size;
	Common::File *f = _vm->resource()->findSound(name, &size);
	if (f) {
		playSoundData(f, size, isSpeech ? &_speechHandle : &_sfxHandle);
		_speechSfxExists = isSpeech;
	} else {
		_speechSfxExists = false;
	}
}

void SBSound::playSoundData(Common::File *f, uint32 size, Audio::SoundHandle *soundHandle) {
	// In order to simplify the code, we don't parse the .sb header but hard-code the
	// values. Refer to tracker item #1876741 for details on the format/fields.
	int headerSize;
	f->seek(2, SEEK_CUR);
	uint16 version = f->readUint16LE();
	switch (version) {
	case 104:
		headerSize = SB_HEADER_SIZE_V104;
		break;
	case 110:
		headerSize = SB_HEADER_SIZE_V110;
		break;
	default:
		warning("Unhandled SB file version %d, defaulting to 104", version);
		headerSize = SB_HEADER_SIZE_V104;
		break;
	}
	f->seek(headerSize - 4, SEEK_CUR);
	size -= headerSize;
	uint8 *sound = (uint8 *)malloc(size);
	if (sound) {
		f->read(sound, size);
		Audio::Mixer::SoundType type = (soundHandle == &_speechHandle) ? Audio::Mixer::kSpeechSoundType : Audio::Mixer::kSFXSoundType;

		Audio::AudioStream *stream = Audio::makeRawStream(sound, size, 11840, Audio::FLAG_UNSIGNED);
		_mixer->playStream(type, soundHandle, stream);
	}
}

AmigaSound::AmigaSound(Audio::Mixer *mixer, QueenEngine *vm)
	: Sound(mixer, vm), _fanfareRestore(0), _fanfareCount(0), _fluteCount(0) {
}

void AmigaSound::playSfx(uint16 sfx) {
	if (_vm->logic()->currentRoom() == 111) {
		// lightning sound
		playSound("88SSSSSS");
	}
}

void AmigaSound::playSong(int16 song) {
	debug(2, "Sound::playSong %d override %d", song, _lastOverride);

	if (song < 0) {
		stopSong();
		return;
	}

	// remap song numbers for the Amiga
	switch (song) {
	case 1:
	case 2:
		song = 39;
		break;
	case 37:
	case 52:
	case 196:
		song = 90;
		break;
	case 38:
	case 89:
		song = 3;
		break;
	case 24:
	case 158:
		song = 117;
		break;
	case 71:
	case 72:
	case 73:
	case 75:
		song = 133;
		break;
	case 203:
		song = 67;
		break;
	case 145:
		song = 140;
		break;
	case 53:
	case 204:
		song = 44;
		break;
	case 136:
	case 142:
	case 179:
		song = 86;
		break;
	case 101:
	case 102:
	case 143:
		song = 188;
		break;
	case 65:
	case 62:
		song = 69;
		break;
	case 118:
	case 119:
		song = 137;
		break;
	case 130:
	case 131:
		song = 59;
		break;
	case 174:
	case 175:
		song = 57;
		break;
	case 171:
	case 121:
		song = 137;
		break;
	case 138:
	case 170:
	case 149:
		song = 28;
		break;
	case 122:
	case 180:
	case 83:
	case 98:
		song = 83;
		break;
	case 20:
	case 33:
		song = 34;
		break;
	case 29:
	case 35:
		song = 36;
		break;
	case 7:
	case 9:
	case 10:
		song = 11;
		break;
	case 110:
		song = 94;
		break;
	case 111:
		song = 95;
		break;
	case 30:
		song = 43;
		break;
	case 76:
		song = 27;
		break;
	case 194:
	case 195:
		song = 32;
		break;
	}

	if (_lastOverride != 32 && _lastOverride != 44) {
		if (playSpecialSfx(song)) {
			return;
		}
	}

	if (_lastOverride == song && _mixer->isSoundHandleActive(_modHandle)) {
		return;
	}
	switch (song) {
	// hotel
	case 39:
		playModule("HOTEL", 1);
		break;
	case 19:
		playModule("HOTEL", 3);
		break;
	case 34:
		playModule("HOTEL", 2);
		break;
	case 36:
		playModule("HOTEL", 4);
		_fanfareRestore = _lastOverride;
		_fanfareCount = 60;
		break;
	// jungle
	case 40:
		playModule("JUNG", 1);
		_fanfareRestore = _lastOverride;
		_fanfareCount = 80;
		_fluteCount = 100;
		break;
	case 3:
		playModule("JUNG", 2);
		_fluteCount = 100;
		break;
	// temple
	case 54:
		playModule("TEMPLE", 1);
		break;
	case 12:
		playModule("TEMPLE", 2);
		break;
	case 11:
		playModule("TEMPLE", 3);
		break;
	case 31:
		playModule("TEMPLE", 4);
		_fanfareRestore = _lastOverride;
		_fanfareCount = 80;
		break;
	// floda
	case 41:
		playModule("FLODA", 4);
		_fanfareRestore = _lastOverride;
		_fanfareCount = 60;
		break;
	case 13:
		playModule("FLODA", 3);
		break;
	case 16:
		playModule("FLODA", 1);
		break;
	case 17:
		playModule("FLODA", 2);
		break;
	case 43:
		playModule("FLODA", 5);
		break;
	// end credits
	case 67:
		playModule("TITLE", 1);
		break;
	// intro credits
	case 88:
		playModule("TITLE", 1);
		break;
	// valley
	case 90:
		playModule("AWESTRUK", 1);
		break;
	// confrontation
	case 91:
		playModule("'JUNGLE'", 1);
		break;
	// Frank
	case 46:
		playModule("FRANK", 1);
		break;
	// trader bob
	case 6:
		playModule("BOB", 1);
		break;
	// azura
	case 44:
		playModule("AZURA", 1);
		break;
	// amazon fortress
	case 21:
		playModule("FORT", 1);
		break;
	// rocket
	case 32:
		playModule("ROCKET", 1);
		break;
	// robot
	case 92:
		playModule("ROBOT", 1);
		break;
	default:
		// song not available in the amiga version
		return;
	}
	_lastOverride = song;
}

void AmigaSound::stopSfx() {
	_mixer->stopHandle(_sfxHandle);
}

void AmigaSound::stopSong() {
	_mixer->stopHandle(_modHandle);
	_fanfareCount = _fluteCount = 0;
}

void AmigaSound::updateMusic() {
	if (_fanfareCount > 0) {
		--_fanfareCount;
		if (_fanfareCount == 0) {
			playSong(_fanfareRestore);
		}
	}
	if (_fluteCount > 0 && (_lastOverride == 40 || _lastOverride == 3)) {
		--_fluteCount;
		if (_fluteCount == 0) {
			playPattern("JUNG", 5 + _vm->randomizer.getRandomNumber(6));
			_fluteCount = 100;
		}
	}
}

void AmigaSound::playSound(const char *base) {
	debug(7, "AmigaSound::playSound(%s)", base);
	char soundName[20];
	sprintf(soundName, "%s.AMR", base);

	uint32 soundSize;
	Common::File *f = _vm->resource()->findSound(soundName, &soundSize);
	if (f) {
		uint8 *soundData = (uint8 *)malloc(soundSize);
		if (soundData) {
			f->read(soundData, soundSize);

			Audio::AudioStream *stream = Audio::makeRawStream(soundData, soundSize, 11025, 0);
			_mixer->playStream(Audio::Mixer::kSFXSoundType, &_sfxHandle, stream);
		}
	}
}

Audio::AudioStream *AmigaSound::loadModule(const char *base, int num) {
	debug(7, "AmigaSound::loadModule(%s, %d)", base, num);
	char name[20];

	// load song/pattern data
	uint32 sngDataSize;
	sprintf(name, "%s.SNG", base);
	uint8 *sngData = _vm->resource()->loadFile(name, 0, &sngDataSize);
	Common::MemoryReadStream sngStr(sngData, sngDataSize);

	// load instruments/wave data
	uint32 insDataSize;
	sprintf(name, "%s.INS", base);
	uint8 *insData = _vm->resource()->loadFile(name, 0, &insDataSize);
	Common::MemoryReadStream insStr(insData, insDataSize);

	Audio::AudioStream *stream = Audio::makeRjp1Stream(&sngStr, &insStr, num, _mixer->getOutputRate());

	delete[] sngData;
	delete[] insData;

	return stream;
}

void AmigaSound::playModule(const char *base, int song) {
	_mixer->stopHandle(_modHandle);
	Audio::AudioStream *stream = loadModule(base, song);
	if (stream) {
		_mixer->playStream(Audio::Mixer::kMusicSoundType, &_modHandle, stream);
	}
	_fanfareCount = 0;
}

void AmigaSound::playPattern(const char *base, int pattern) {
	_mixer->stopHandle(_patHandle);
	Audio::AudioStream *stream = loadModule(base, -pattern);
	if (stream) {
		_mixer->playStream(Audio::Mixer::kSFXSoundType, &_patHandle, stream);
	}
}

bool AmigaSound::playSpecialSfx(int16 sfx) {
	switch (sfx) {
	case 5: // normal volume
		break;
	case 15: // soft volume
		break;
	case 14: // medium volume
		break;
	case 25: // open door
		playSound("116BSSSS");
		break;
	case 26: // close door
		playSound("105ASSSS");
		break;
	case 56: // light switch
		playSound("27SSSSSS");
		break;
	case 57: // hydraulic doors open
		playSound("96SSSSSS");
		break;
	case 58: // hydraulic doors close
		playSound("97SSSSSS");
		break;
	case 59: // metallic door slams
		playSound("105SSSSS");
		break;
	case 63: // oracle rezzes in
		playSound("132SSSSS");
		break;
	case 27: // cloth slide 1
		playSound("135SSSSS");
		break;
	case 83: // splash
		playSound("18SSSSSS");
		break;
	case 85: // agression enhancer
		playSound("138BSSSS");
		break;
	case 68: // dino ray
		playSound("138SSSSS");
		break;
	case 140: // dino transformation
		playSound("55BSSSSS");
		break;
	case 141: // experimental laser
		playSound("55SSSSSS");
		break;
	case 94: // plane hatch open
		playSound("3SSSSSSS");
		break;
	case 95: // plane hatch close
		playSound("4SSSSSSS");
		break;
	case 117: // oracle rezzes out
		playSound("70SSSSSS");
		break;
	case 124: // dino horn
		playSound("103SSSSS");
		break;
	case 127: // punch
		playSound("128SSSSS");
		break;
	case 128: // body hits ground
		playSound("129SSSSS");
		break;
	case 137: // explosion
		playSound("88SSSSSS");
		break;
	case 86: // stone door grind 1
		playSound("1001SSSS");
		break;
	case 188: // stone door grind 2
		playSound("1002SSSS");
		break;
	case 28: // cloth slide 2
		playSound("1005SSSS");
		break;
	case 151: // rattle bars
		playSound("115SSSSS");
		break;
	case 152: // door dissolves
		playSound("56SSSSSS");
		break;
	case 153: // altar slides
		playSound("85SSSSSS");
		break;
	case 166 : // pull lever
		playSound("1008SSSS");
		break;
	case 182: // zap Frank
		playSound("1023SSSS");
		break;
	case 69: // splorch
		playSound("137ASSSS");
		break;
	case 70: // robot laser
		playSound("61SSSSSS");
		break;
	case 133: // pick hits stone
		playSound("71SSSSSS");
		break;
	case 165: // press button
		playSound("1007SSSS");
		break;
	default:
		return false;
	}
	return true;
}

} // End of namespace Queen
